training/src/main/scala/com/airbnb/aerosolve/training/DecisionTreeTrainer.scala

```diff
@@ -129,94 +167,172 @@ object DecisionTreeTrainer {
+      if (response) {
+        if (label) {
+          rightPos += 1.0
+        } else {
+          rightNeg += 1.0
+        }
+      } else {
+        if (label) {
+          leftPos += 1.0
+        } else {
+          leftNeg += 1.0
+        }
+      }
+    }
+    val rightCount = rightPos + rightNeg
+    val leftCount = leftPos + leftNeg
+
+    if (rightCount >= minLeafCount && leftCount >= minLeafCount) {
+      val p1 = rightPos / rightCount
+      val n1 = rightNeg / rightCount
+      val f1 = rightCount / (leftCount + rightCount)
+      val p2 = leftPos / leftCount
+      val n2 = leftNeg / leftCount
+      val f2 = leftCount / (leftCount + rightCount)
+      splitCriteria match {
+        case "gini" => {
+          // Using negative gini since we are maximizing.
+          val gini = -(
+              f1 * (p1 * (1.0 - p1) + n1 * (1.0 - n1)) +
+                  f2 * (n2 * (1.0 - n2) + p2 * (1.0 - p2))
+              )
+
+          Some(gini)
+        }
+        case "information_gain" => {
+          var ig = 0.0
+          if (p1 > 0) {
+            ig += f1 * p1 * scala.math.log(p1)
+          }
+          if (n1 > 0) {
+            ig += f1 * n1 * scala.math.log(n1)
+          }
+          if (p2 > 0) {
+            ig += f2 * p2 * scala.math.log(p2)
+          }
+          if (n2 > 0) {
+            ig += f2 * n2 * scala.math.log(n2)
+          }
+
+          Some(ig)
+        }
+        case "hellinger" => {
+          val scale = 1.0 / (leftCount * rightCount)
+          // http://en.wikipedia.org/wiki/Bhattacharyya_distance
+          val bhattacharyya = math.sqrt(leftPos * rightPos * scale) + math.sqrt(leftNeg * rightNeg * scale)
+          // http://en.wikipedia.org/wiki/Hellinger_distance
+          val hellinger = math.sqrt(1.0 - bhattacharyya)
+
+          Some(hellinger)
+        }
+      }
+    } else {
+      None
+    }
+  }
+
+  // Evaluate a regression-type split
+  // See http://www.stat.cmu.edu/~cshalizi/350-2006/lecture-10.pdf for overview of algorithm used
+  def evaluateRegressionSplit(
+      examples : Array[util.Map[java.lang.String, util.Map[java.lang.String, java.lang.Double]]],
+      rankKey : String,
+      minLeafCount : Int,
+      splitCriteria : String,
+      candidateOpt : Option[ModelRecord]): Option[Double] = {
+    var rightCount: Double = 0.0
+    var rightMean: Double = 0.0
+    var rightSumSq: Double = 0.0
+    var leftCount: Double = 0.0
+    var leftMean: Double = 0.0
+    var leftSumSq: Double = 0.0
+
+    for (example <- examples) {
+      val response = BoostedStumpsModel.getStumpResponse(candidateOpt.get, example)
+      val labelValue = example.get(rankKey).asScala.head._2
+
+      // Using "online algorithm" for computing mean and sum-squared errors as
+      // described in https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance
+      if (response) {
+        rightCount += 1
+        val delta = labelValue - rightMean
+        rightMean += delta / rightCount
```

Comment:
This is not the right way to compute the variance .. you should accumulate the sum and sum squared and only divide by the count at the very end outside of the loop